### DISCUSSION TITLE
Overcommitting storage silently corrupts data

### DISCUSSION THREAD

**Discussion by crackerjam** [stratis-storage/project#652] [Q&A]
Trying to copy a 10GB file to a 2GB Stratis filesystem appears to succeed, but in reality causes the file to be corrupted. To reproduce:

1. Login to a RHEL 9.2 system as root (System has a 20GB root disk, plus a 2GB blank disk)
2. Install/enable/start `stratisd` and `stratis-cli` (version 3.4.1)
`yum -y install stratisd stratis-cli; systemctl enable --now stratisd`
3. `stratis pool create testpool /dev/sdb`
4. `stratis filesystem create testpool testfilesystem`
5. `mkdir -p /mnt/test`
6. Add this line to `/etc/fstab`
`/dev/stratis/testpool/testfilesystem    /mnt/test       xfs     defaults,x-systemd.requires=stratisd.service 0 0`
7. `mount -a`
8. Download a large file to /root for testing
`wget -P /root http://mirrors.rit.edu/centos/7.9.2009/isos/x86_64/CentOS-7-x86_64-Everything-2009.iso`
9. Copy the file to the Stratis filesystem
`cp /root/CentOS-7-x86_64-Everything-2009.iso /mnt/test`
10. Hash the test file in both locations
`md5sum /root/CentOS-7-x86_64-Everything-2009.iso`
`md5sum /mnt/test/CentOS-7-x86_64-Everything-2009.iso`

These checksums will come back different, indicating data corruption.

**Comment by mulkieran** [comment]
Thanks for the report. We believe that this problem has been addressed in subsequent stratisd releases, which are available in RHEL 9. Here is some documentation on the design changes that we made in Stratis 3 to make extension of the Stratis filesystems more robust: https://stratis-storage.github.io/thin-provisioning-redesign/ . If you  reproduce the same problem in a more recent version of Stratis, please let us know, we will be eager to investigate. Thanks!

**Comment by crackerjam** [comment]
@mulkieran This is NOT fixed on RHEL 9.2 with stratis 3.4.1. I just re-ran the steps in my original post on a fresh system and see the same behavior.

I've updated my original post to include those versions.

**Comment by mulkieran** [comment]
@crackerjam Can you confirm that ```/dev/sdb``` is the 2 GiB device that you have in your initial statement? I may have initially misread your statement, and thought that the device you were using for the Stratis pool was the 20 GiB one that you mention first. If that is the case, i.e., that the pool itself is only 2 GiB there is not much to do, because the cp command itself does not perform any sync operation or confirm that the write has completed.

The pool itself is supposed to maintain information about itself, and can give some warnings. When you first create the pool, it will assess its state and may issue a warning, which will be seen with the ```stratis pool list``` command. After you issue the cp command, the pool will try to accommodate the additional data by expanding the filesystem. If you list the pool after the failed cp command, you may see additional warnings.

**Comment by crackerjam** [comment]
@mulkieran Correct, /dev/sdb is a 2GiB disk.

> If that is the case, i.e., that the pool itself is only 2 GiB there is not much to do, because the cp command itself does not perform any sync operation or confirm that the write has completed.

That doesn't seem to be correct. If I just format the raw disk with XFS directly and try to run the copy, I receive:

`cp: error writing '/mnt/test/CentOS-7-x86_64-Everything-2009.iso': No space left on device`

> If you list the pool after the failed cp command, you may see additional warnings.

`stratis pool list` does show an alert of "WS0001" next to the pool, which is something, but honestly not very helpful. Googling "stratis WS0001" doesn't return any results, so I'm not even able to identify what this error code means.


Think of a scenario where you have a stratis volume that receives and stores critical business documents. Sure, ideally you would have monitoring set up for disk usage, but let's say you don't. With any other filesystem your application would start receiving errors when the disk is full. Users will see that something's going on, it will be clear that things aren't working properly. Things will grind to a halt, but you won't lose data.

With this issue, everything just appears to silently work. Files show up on disk and report correct sizes. However, when you read the files, they're junk. Depending on the use case you could go weeks or months without noticing that all of your data is just evaporating. This should be an extremely critical priority issue.

**Comment by mulkieran** [comment]
@crackerjam In a case like this, you would be advised to create a pool with no overprovisioning mode set. In such a pool the size of the XFS filesystem can not be made any larger that the actual space that is available to it. In that case, the filesystem will behave like the one that you placed on the raw device.

For the raw device, check the filesystem size. You'll notice that it is significantly smaller than the default filesystem size, 1 TiB, that Stratis creates when overprovisioning is allowed.

**Comment by crackerjam** [comment]
@mulkieran I don't think that's really an appropriate stance. What is 'a case like this'? A case where you don't want your files to get corrupted? No filesystem, ever, should report back that a write succeeded when it did not, regardless of options used on creation. That's just a core tenant of how filesystems work. It is especially true here, when I'm using the default options.


I have to be honest, I only discovered Stratis as part of some RHEL training. Personally, and professionally, if I ever get a whiff of anyone thinking about using Stratis I will tell them to steer clear due to how this bug is being handled. In my opinion, this is a massive issue that needs to be resolved if you want this to be taken seriously as a replacement for LVM, and I can only imagine how many other data-destroying bugs there are that have been ignored because "well you should have just used the right command arguments".

If you'd like any more assistance testing scenarios around this, please feel free to let me know. Otherwise, it seems like you have a particular opinion here that I probably won't be able to change.

**Comment by mulkieran** [comment]
@crackerjam I'm afraid that you are failing to understand how filesystems work. To produce an analogous situation with LVM, simply create a thin pool on your 2 GiB device, construct a thin device of say, 1 TiB, on that thinpool, create an XFS filesystem on it, and do your cp as before. The cp will return success, but the file will not be copied in toto.

It is surprising to some people that that is how filesystems behave, but it is a consequence of the original design choices made for filesystems in a simple world that did not include thinly provisioned devices supplied by kernel modules.

If you want to do a copy that is sure to fail if there is really no place to write the bytes, you will have to explicitly sync your data.

**Reply by crackerjam** [reply]
Okay, I was able to replicate that with LVM and see what you're saying. However, it took a bit to get there and I continue to have some gripes.

1. This is not default behavior with LVM. You need to create a thin volume inside a thin pool in order get this to work, along with specifying that the volume is larger than the pool it goes into, which isn't really a standard configuration. It has uses, sure, but I wouldn't say that it's common. 
2. When you create volumes like this LVM produces a big warning message alerting that this may break things.

I still think, ideally, stratis volumes should report write failures back up the stack when situations like this occur. If the goal is to be better than LVM this would be a place open for improvement. However, if that isn't possible there should at least be feedback when creating filesystems identifying this potential issue. The behavior is just so far outside of what 99% of Linux engineers will consider normal it really should be called out.

**Reply by mulkieran** [reply]
@crackerjam The reason the default size for the filesystem is 1 TiB is to avoid a performance penalty that results when an XFS filesystem is grown from a small size to a larger size. The general rule is that an XFS filesystem should not be increased in size by more than a factor of 8, in order to avoid this performance penalty. Stratis has introduced a new filesystem limits functionality in the most recent release[1]. So, as an administrator you can set that limit and when the filesystem reaches that limit it can not be grown and XFS will reject writes when the filesystem is out of space. However, the limit can not be less than the current filesystem size, so this would require planning when the filesystem is created.

You're correct that it would be ideal for the out-of-space condition on the thin device to be tidily propagated through the filesystem layer to the user program. But that requires a level of communication between the filesystem and the block layer which has yet to be achieved.

[1] https://stratis-storage.github.io/stratis-release-notes-3-6-0/

**Comment by yyshell** [comment]
Now on RHEL 9.3, the problem is still not fixed.  and if we use the "no-overprovision" option. the file system size can not grow any more. This is a very damning point. It's almost impossible for me to use Stratis in a production environment.

**Reply by jbaublitz** [reply]
If you are seeing that overprovisioning is causing your filesystem to stop growing, that is a sign that you have exhausted your physical space and need to add more physical storage.

**Comment by wushilin** [comment]
Well I am surprised to see this, but it is indeed still an issue with 3.6.8 daemon and 3.6.2 CLI.

The file system is not corrupted, just that the file that apparently to be copied successfully, was truncated or zero sized.

**Reply by jbaublitz** [reply]
Currently, we are having discussions with the XFS and thin provisioning team about whether we would be able to get them to return ENOSPC on writes that would exhaust physical space. I'm leading the discussion, so I'll pass along that there's someone else expressing interest in this. This exact piece is out of Stratis's hands, and while we can advocate for different behavior, this is not something we can avoid short of providing the no-overprovisioning option that we already have to ensure that this doesn't happen. fsync still is the best way to see if your write operations have succeeded on any filesystem in Linux or whether it failed on writeback, but I will let you know if there's any motion on the XFS/thin provisioning integration.

**Comment by jbaublitz** [comment]
@crackerjam We've been discussing this with the filesystem and block storage teams and I'm wondering whether for your use case you would prefer if we remounted the filesystems as read-only to prevent data from being written that can't be persisted due to the thin pool running out of space. The general feedback we've gotten from both teams is that they are not confident that there is a solution that will be accepted upstream in the kernel due to things like performance when implementing the communication between XFS and the thin pool.

**Reply by crackerjam** [reply]
Thanks to you and the team for keeping this issue on your radar.

I think that may be a fine resolution depending on how applications receive information on those events. If a file can't be persisted because of the physical volume running out of space the application must receive some sort of write failure notification.

For example, if you write a file with `cp` that doesn't fit, `cp` should error with a write failure immediately. If that bad write succeeds but now the volume is read-only so future writes can't happen, that is still better than the existing system, but I would still say it is not acceptable for a production filesystem. Data should never be corrupted from normal activity.

**Reply by jbaublitz** [reply]
I agree that data should never be corrupted from normal activity, but I will also just highlight that this is not corruption from normal activity per-se. This is a situation where data is persisted to the cache successfully and then when the cache is flushed, the write does indeed fail, but because this is done asynchronously, it is in the logs but not provided directly to the user on the write call. Now there are ways that the kernel team could get XFS to check whether thin pools have enough space, but the end result of this would likely be a reduction in performance. Given that a main selling point of thin provisioning is performance and flexibility, this would likely not be accepted upstream or desirable for a lot of users if they experienced degraded performance as a result of this. I do also want to highlight that there is also a very doable workaround for this which may not be desirable for performance reasons, but per the Linux filesystem API design, if you call fsync after writes are performed, you will indeed get confirmation of whether that write succeeded or not. The Linux filesystem API is not designed to guarantee that data has been persisted in Stratis or any other storage stack even if the write succeeds. That is handled by the return code of fsync. Just to drive the few points here home, it would be entirely doable for us to naively put some sort of kernel based layer on top of the filesystem to drive an fsync for every write that comes through that layer and this would give you the desired user experience of writes reporting whether the data has been persisted or not for each write. The tradeoff here is there would likely be a significant performance degradation from this change as the filesystem does a lot of work before writeback to minimize unnecessary/inefficient IO. The performance would likely be even worse than an integration between XFS and the thin pool but even an integration between XFS and the thin pool has been tried a number of times without finding a solution that did not include tradeoffs that were not acceptable to the wider community. While I do want to find a way to make this work, I would argue that rather any application that does not call fsync to confirm that data has been successfully persisted to storage is broken and could lose data at any time.

I do want to find a solution that works for you. I understand that this user experience might be surprising, confusing, and generally counter intuitive given that some writes will return a failure if inability to persist is able to be detected at that time. That being said, I think it's important to understand that some of this is kernel API design and while LVM and Stratis can provide feedback to the kernel teams working on this, this would be a massive departure from how the kernel has done things historically and would likely need a more comprehensive solution than I think anyone has proposed so far. It would also take a lot of effort (and likely even more time given that the kernel side is more cautious to make large changes like this) to convince the upstream teams to redesign an API simply for this use case.

We are willing to implement features to help users more easily more easily manage their storage. However, when we tried to provide a solution to report immediate information on whether the write succeeded or not due to space constraints (no overprovisioning mode), users did not find that helpful because they could not grow their filesystems past the physical limitations of the storage device on which it was based. We're open to implementing what we can but want to ensure that it's helpful to users before doing so.

**Reply by crackerjam** [reply]
Gotcha, thanks for the in-depth explanation. I think having the filesystem go read-only when the thin pool is full is probably the best solution considering the tradeoffs of a more robust implementation.

**Reply by jbaublitz** [reply]
@crackerjam So after having an in-depth discussion with the XFS team, it appears at the kernel level we will not be able to remount as read only. As long as there is an open write file descriptor, we'll get EBUSY if we try to remount as read-only. This is likely to occur in any production use case. However, one thing we can do is use `xfs_freeze`. Would this be suitable? My understanding of the behavior is that any writer that is already open will block on writes so data cannot be written until more data space is added to the pool. New writers will fail to open. This seems like the best we can do so would this be helpful to you?

**Reply by crackerjam** [reply]
@jbaublitz I think that would be great, yes. Thank you again.

**Reply by jbaublitz** [reply]
@crackerjam We've been discussing this for a very long time on the team. I've proposed a number of solutions and each one seems to come with a lot of drawbacks.

1. Mount the filesystem as read-only: this is not possible and will fail with EBUSY if there are any writers open on the filesystem.
2. xfs_freeze: this requires flushing all of the data already in the page cache. This may result in lost data if the underlying thin layer does not have enough space to receive all of the data in the page cache. Predicting the threshold where we should flush is virtually impossible without overestimating significantly, possibly cutting off writes with a good amount of free space left.

Someone on the XFS team recently proposed a solution and this one seems to be blessed by the XFS team as potential way around this problem.

The idea would be to keep retrying writeback from XFS into the thin pool until it succeeds. This means that you would be required to add more space to the pool if you receive a particular D-Bus signal indicating that the thin pool is out of space. If you add it, the data will be persisted once the new space is available. I'm curious what your thoughts are on this particular point. It would require you to add more space but would prevent the current behavior of losing that data if you add more space when you receive the signal.

**Reply by jbaublitz** [reply]
Just one last note: it would also be possible to resolve the problem by freeing up space on the filesystem/thin pool.

**Reply by crackerjam** [reply]
@jbaublitz Thanks for reaching out, I appreciate that this is still on the team's radar. I think my question on that would be how would that actually look to a user on the OS? For a simple example, if I'm running `cp` to a filesystem that runs out of space like this and that retry mechanism takes place, how does that look to the end user?

**Reply by jbaublitz** [reply]
@crackerjam Thank you for your patience. As the discussion was still ongoing, I wanted to tie up a few loose ends. I have a prototype that I have tested that has the following properties.

1. When the file is copied it does not disappear if there is not enough room on the pool *unless* the user stops the pool, shuts down the system, or a similar action before persisting the file. This we can definitely not do anything about as we cannot provide the user with sufficient space to persist data. Calling sync will indefinitely hang which should be an indication that not all data can be persisted/something is blocking writeback.
2. A user is able to free up space on the filesystem or add more space to the pool to enable writeback to progress. Adding more space is very straightforward. Freeing up space on the filesystem may require fstrim or some other instruction to the thinpool to free up the space that was reclaimed on the filesystem. This is relatively standard for thin provisioning and we're actively investigating ways to smooth out this process for the user.
3. Performance of xfs will be relatively unaffected. We have also been investigating a passthrough mode for XFS and while this will provide the user with information at the time of the write operation about whether there is enough space, this has the potential to bypass some optimizations that XFS makes and could result in worse performance.

Does this sound like a workable solution to you? I think it resolves the problem of the data disappearing on writeback (assuming you eventually supply enough space to persist the data before shutdown) and gives a few flexible options for providing that space.

**Reply by crackerjam** [reply]
@jbaublitz I'm thinking through the reality of this and it definitely has some benefits. The behavior will bubble up through the stack and present obvious issues to applications failing to write, which will end at some human person noticing that things are broken. This solves the original problem.

My only hesitance is that, in every company I've been in, when the level 1 operations team gets a ticket that a server isn't working right, the absolute first thing they do is reboot it. There are already system types where this is messy (e.g. databases), but that doesn't change it being the go-to process for most teams.

Would it be impossible for stratis to enforce some sort of 'emergency zone' where it reserves a bit of space specifically for this problem? If you're hanging writes anyway, I would think you could just have a very small reserved area for the cached writes to sit so they can survive a hasty reboot. On reboot, stratis can just check for free space, if there's none, go read only and keep that cache preserved for whenever there is space again.

**Reply by jbaublitz** [reply]
@crackerjam Just to make sure we're on the same page, the two options we're considering are:
1. passthrough mode - We have done less research into this. This will result in write errors provided to the user when no space is available but could negatively impact performance.
2. queuing mode - This keeps the writes in memory until space is freed up or added to persist it. Discards/deletes are handled immediately to allow freeing up space.

I have a few concerns with your last message. I think we need to divide up concerns into desktop use cases and enterprise use cases. For desktop use cases, it's entirely possible that users will not call fsync to check that the data has persisted and the hope is that the data will eventually make its way to writeback if enough space is available. For these kinds of workflows, that may be good enough. However for enterprise customers, the official way to validate that your data has been persisted to disk is to call fsync. There is no way around this. I have talked with a large number of storage maintainers in the ecosystem about this particular problem. Unless you call fsync after writing the data, even without thin provisioning, you will not get a status of whether the IO was persisted and your data could very well not be there when you check for it. Power outages occurring before writeback is a great example of this. You were saying that the problem is that if the write succeeds, you expect to see the data there. If you were to do the cp command you originally used in your description of the problem and then call fsync, you would indeed get a report that the write failed.

I also wanted to clear up what I think is a misunderstanding regarding the solution I'm proposing. As I previously mentioned, switching to read-only requires the filesystem having no writers open on the filesystem or remounting as read-only will return `EBUSY`. Our only real option would be to go around sending `SIGTERM`/`SIGKILL` to processes with open writable file descriptors on the filesystem to remount as read-only. This would make a lot of people very unhappy when their apps get killed by the filesystem so I believe remounting as read-only can be ruled out for this particular scenario. Similarly, freezing all writers requires flushing all data in the cache to disk before writers can be frozen. This presents a problem because if there is not enough space to receive all of the data that is being flushed to disk, we're back to square one: data disappearing during writeback. The XFS team has strongly recommended against this approach as it is impossible to estimate the amount of space that will be required in the thin pool to persist the cached IO. The most recent solution I've proposed will provide the following: when data fails on writeback, it remains queued until more space becomes available either through adding a disk (which is something an enterprise customer would be equipped to do) or freeing up space on the disk (something a desktop user should be familiar with). This will at the very least address your original concern about failures on writeback by allowing time to provide/free up that extra space without the writeback failing and the data being lost.

I also wanted to respond to your idea about keeping emergency space for these kinds of situations. This is not a workable solution in all or even most cases. We bump into the same problem as freezing writers: estimating the amount of space to reserve for rescue operations is impossible. XFS provides no way to estimate how many blocks will be consumed in the thin pool on writeback. The end result is that this solution would sometimes work and sometimes it would fail and result in failed writeback providing inconsistent improvement. I think this will end up mostly making everyone unhappy as it will be completely unpredictable when it will and won't work.

Now as for suggestions about what users can do to avoid this, we provide a mode in Stratis called no overprovisioning. This mode ensures that the total filesystem/snapshot size is never larger than the physical space available. This will inherently eliminate this problem. Users can also call fsync on writes to check the status of their writes. This will notify the user if their data failed to persist to disk. The final option of indefinitely queuing requests until space is freed up or added would be mostly a quality of life improvement. The goal here is to avoid data writeback failing in cases where the user can quickly respond and free up/add more space. Outside of this, I don't believe there is a lot we can do.

I want to highlight that this discussion started from a problem statement of writes being accepted but data not being persisted without calling fsync. While I am open to trying to find solutions that improve the user's experience, if you are not calling fsync and still expect your data to be there, you are misusing the storage API.

**Reply by crackerjam** [reply]
@jbaublitz Thanks for the write up. To be completely clear, my concerns are !00% on the enterprise side. I am a senior infrastructure engineer for a large enterprise, so that's the perspective I'm speaking from. I know I'm mentioning errors becoming obvious to humans at some point, but I'm talking about the process of a system failing, an alert being generated in a monitoring application, and someone getting a call about it. I have no opinion on desktop use because I do not use or support any desktop Linux systems.

I think the queueing mode solution will be a good one. I do still wonder about 'emergency mode' though. I don't have a deep understanding of this, so I might be naive here, but I would expect that the writes a process is going to try to make, which would be stalled here, would be at the block level, so 64KB max, right? Maybe multiple processes writing would build up a handful of those at the same time, but with this line of thought I would think a 100MB reservation would provide a more than healthy buffer where most people won't miss the space.

If my understanding is incorrect on what we're persisting there then I retract that suggestion. Though, then that makes me wonder, if we're caching these writes in memory how fast are we going to blow the server out of available RAM?

**Reply by jbaublitz** [reply]
If you are talking about enterprise use cases, then you should be using fsync to validate that data has been persisted, full stop. If you are not, there is no guarantee that the data will be persisted using *or* avoiding thin provisioning and a write being accepted will not tell you if writeback failed along the way down the stack. This behavior is even possible with something as simple as XFS on a hard drive partition.

With regards to your proposed limit on write size, while an individual IO may have an upper limit, the amount of data that the filesystem can accept is the amount of free space on the filesystem (sometimes as large as 1TiB even if you only have a few GiBs of physical space available). This is why I really want to emphasize that if you're looking for the filesystem to return ENOSPC, I would strongly recommend the no overprovisioning mode. This sets the upper limit of the total filesystem size to be equal to the total space available. If your filesystem is 1 TiB and your physical storage is 100GiB, the upper limit of the writes the filesystem can accept is basically what you're able to fit in the page cache with a hard limit of 1TiB (the filesystem size).

However, if you want to use thin provisioning and overprovision the storage, this comes with a tradeoff. The tradeoff is that the block layer and filesystem layer are operating completely independently and while thin provisioning provides a lot of great features like efficient snapshots, the drawback is that you have to carefully manage out-of-space conditions.

Just to give you an idea of what your options are here:
1. Use no overprovisioning.
2. We can implement this queuing mode but it is up to you to respond to the signal we will send and either delete files to make room or add more space to the pool. The writes will not fail on writeback during normal out-of-space conditions but you do have to respond quickly to add more storage. In a cloud environment this can even be automated: listen for the signal that the pool is out of space and add more storage to the pool.
